File: javascript/src/problems1_100/merge-k-sorted-lists.js
import SinglyListNode from '../concepts/singly-list-node';

class Problem {
  mergeKLists(lists) {
    throw new Error(`Not implemented (${lists}).`);
  }
}

class Default extends Problem {
  mergeKLists(lists) {
    const nodes = lists.filter(n => n);
    if (nodes.length === 0) {
      return undefined;
    }

    const root = new SinglyListNode();
    let current = root;
    while (nodes.length > 0) {
      // find lowest
      const lowest =
          nodes.reduce((previous, current) => current.val < previous.val ? current : previous);

      // append lowest to current
      current.next = new SinglyListNode(lowest.val);
      current = current.next;

      // remove the lowest node within list
      if (!lowest.next) {
        nodes.splice(nodes.indexOf(lowest), 1);
        continue;
      }
      const nxt = lowest.next;
      lowest.val = nxt.val;
      lowest.next = nxt.next;
    }
    return root.next;
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;

File: python/src/problems1_100/merge_k_sorted_lists.py
from abc import ABC, abstractmethod
from typing import override

from src.concepts.singly_list_node import SinglyListNode, SINGLY_LIST_NODE_MAX


class Problem(ABC):
    @abstractmethod
    def merge_k_lists(self, lists: list[SinglyListNode]) -> SinglyListNode | None:
        pass


class Default(Problem):
    @override
    def merge_k_lists(self, lists: list[SinglyListNode]) -> SinglyListNode | None:
        nodes: list[SinglyListNode] = [n for n in lists if n]
        if len(nodes) == 0:
            return None

        root: SinglyListNode = SinglyListNode()
        current: SinglyListNode = root
        while len(nodes) > 0:
            # find lowest
            lowest: SinglyListNode = min(nodes, key=lambda n: n.val, default=SINGLY_LIST_NODE_MAX)

            # append lowest to current
            current.next = SinglyListNode(lowest.val)
            current = current.next

            # remove the lowest node within list
            if not lowest.next:
                nodes.remove(lowest)
                continue
            nxt: SinglyListNode = lowest.next
            lowest.val = nxt.val
            lowest.next = nxt.next
        return root.next


SOLUTIONS: list[Problem] = [Default()]
